pkg/nimble/netif/nimble_netif.c

```diff
@@ -0,0 +1,630 @@
+#define NETTYPE                 GNRC_NETTYPE_UNDEF
+#endif
+
+/* buffer configuration
+ * - we need one RX and one TX buffer per connection */
+#define MTU_SIZE                (NIMBLE_NETIF_MTU)
+#define MBUF_OVHD               (sizeof(struct os_mbuf) + \
+                                 sizeof(struct os_mbuf_pkthdr))
+#define MBUF_SIZE               (MBUF_OVHD + MYNEWT_VAL_BLE_L2CAP_COC_MPS)
+#define MBUF_CNT                (MYNEWT_VAL_BLE_MAX_CONNECTIONS * 2 * \
+                                 ((MTU_SIZE + (MBUF_SIZE - 1)) / MBUF_SIZE))
+
+/* thread flag used for signaling transmit readiness */
+#define FLAG_TX_UNSTALLED       (1u << 13)
+
+/* allocate a stack for the netif device */
+static char _stack[THREAD_STACKSIZE_DEFAULT];
+static thread_t *_netif_thread;
+
+/* keep the actual device state */
+static gnrc_netif_t *_nimble_netif = NULL;
+static gnrc_nettype_t _nettype = NETTYPE;
+
+/* keep a reference to the event callback */
+static nimble_netif_eventcb_t _eventcb;
+
+/* allocation of memory for buffering IP packets when handing them to NimBLE */
+static os_membuf_t _mem[OS_MEMPOOL_SIZE(MBUF_CNT, MBUF_SIZE)];
+static struct os_mempool _mem_pool;
+static struct os_mbuf_pool _mbuf_pool;
+
+/* notify the user about state changes for a connection context */
+static void _notify(int handle, nimble_netif_event_t event)
+{
+    if (_eventcb) {
+        _eventcb(handle, event);
+    }
+}
+
+static void _netif_init(gnrc_netif_t *netif)
+{
+    (void)netif;
+
+    /* save the threads context pointer, so we can set its flags */
+    _netif_thread = (thread_t *)thread_get(thread_getpid());
+
+#ifdef MODULE_GNRC_SIXLOWPAN
+    /* we disable fragmentation for this device, as the L2CAP layer takes care
+     * of this */
+    _nimble_netif->sixlo.max_frag_size = 0;
+#endif
+}
+
+static int _send_pkt(nimble_netif_conn_t *conn, gnrc_pktsnip_t *pkt)
+{
+    int res;
+    int num_bytes = 0;
+
+    if (conn == NULL || conn->coc == NULL) {
+        return -ENOTCONN;
+    }
+
+    /* copy the data into a newly allocated mbuf */
+    struct os_mbuf *sdu = os_mbuf_get_pkthdr(&_mbuf_pool, 0);
+    if (sdu == NULL) {
+        return -ENOBUFS;
+    }
+    while (pkt) {
+        res = os_mbuf_append(sdu, pkt->data, pkt->size);
+        if (res != 0) {
+            os_mbuf_free_chain(sdu);
+            return -ENOBUFS;
+        }
+        num_bytes += (int)pkt->size;
+        pkt = pkt->next;
+    }
+
+    /* send packet via the given L2CAP COC */
+    do {
+        res = ble_l2cap_send(conn->coc, sdu);
+        if (res == BLE_HS_EBUSY) {
+            thread_flags_wait_all(FLAG_TX_UNSTALLED);
+        }
+    } while (res == BLE_HS_EBUSY);
+
+    if ((res != 0) && (res != BLE_HS_ESTALLED)) {
+        os_mbuf_free_chain(sdu);
+        return -ENOBUFS;
+    }
+
+    return num_bytes;
+}
+
+static int _netif_send_iter(nimble_netif_conn_t *conn,
+                            int handle, void *arg)
+{
+    (void)handle;
+    _send_pkt(conn, (gnrc_pktsnip_t *)arg);
```

Comment:
It somewhat worries me, that there is no error reporting for the multicast case :-/. Couldn't `nimble_netif_conn_foreach()` return the value `!= 0` when it is aborted which then in turn is returned in [_netif_send](https://github.com/RIOT-OS/RIOT/pull/11578/files#diff-0fddc9bbf97a3532b485226bb4642e85R162)?

Reply:
I fear that the situation is not that simple. The foreach() loop is never aborted, even if one or more calls to `_send_pkt()` do fail. Due to the nature of how we implemented multicast here, I am not even sure what the best behavior would be here: abort sending packets once the first call to `_send_pkt()` fails? Report and error through `_netif_send()`, even if only one out of X calls to `_send_pkt()` failed?

On top: the pre-provisioning of buffer space should actually prevent the buffers running full, so at least the `ENOBUFS` cases in `_send_pkt()` should never be triggered (in theory...).

So for now, I would like to leave the code as is, and I will certainly re-visit (if not re-implement) this once I get to do my experiments on the multicast behavior of IP-over-BLE soon.

Reply:
Ok. Thanks for the insight.